In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC

from xgboost import XGBClassifier

from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

from sklearn.preprocessing import StandardScaler

In [3]:
df = pd.read_csv("wine_quality.csv")

In [4]:
df.head()

country                                        description  points  price  \
0      US  This tremendous 100% varietal wine hails from ...      96  235.0   
1   Spain  Ripe aromas of fig, blackberry and cassis are ...      96  110.0   
2      US  Mac Watson honors the memory of a wine once ma...      96   90.0   
3      US  This spent 20 months in 30% new French oak, an...      96   65.0   
4  France  This is the top wine from La Bégude, named aft...      95   66.0   

              variety  
0  Cabernet Sauvignon  
1       Tinta de Toro  
2     Sauvignon Blanc  
3          Pinot Noir  
4  Provence red blend

In [5]:
# Pick out features X and labels y
X = df.iloc[:,2:4].values
y = df['country']

In [6]:
X, y.head()

(array([[ 96., 235.],
        [ 96., 110.],
        [ 96.,  90.],
        ...,
        [ 91.,  20.],
        [ 90.,  52.],
        [ 90.,  15.]]), 0        US
 1     Spain
 2        US
 3        US
 4    France
 Name: country, dtype: object)

In [81]:
# Split dataset into test and train
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [82]:
print(X_train)
y_train.head()

[[91. 60.]
 [84. 13.]
 [87. 26.]
 ...
 [87. 28.]
 [90. 26.]
 [93. 55.]]


17649      Spain
87040         US
44300         US
101288    France
81544      Italy
Name: country, dtype: object

In [83]:
# Initialise models
LR_model         = LogisticRegression(random_state=0, solver='lbfgs', multi_class='multinomial')
LDA_model        = LinearDiscriminantAnalysis()
KNN_model        = KNeighborsClassifier()
GaussianNB_model = GaussianNB()
DTree_model      = DecisionTreeClassifier()
SVC_model        = LinearSVC(multi_class='ovr')

In [84]:
xgb_model = XGBClassifier(booster='gbtree', objective='multi:softprob', random_state=42, eval_metric="auc", num_class=4)

## Addition of scaling here

In [85]:
X_train

array([[91., 60.],
       [84., 13.],
       [87., 26.],
       ...,
       [87., 28.],
       [90., 26.],
       [93., 55.]])

In [86]:
points = [ i for [i, j] in X_train]
price  = [ j for [i, j] in X_train]

In [87]:
points[0], price[0]

(91.0, 60.0)

In [88]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)

In [89]:
X_train

array([[ 0.92877153,  0.62812109],
       [-1.20512293, -0.58475479],
       [-0.29059673, -0.24927848],
       ...,
       [-0.29059673, -0.19766674],
       [ 0.62392946, -0.24927848],
       [ 1.53845566,  0.49909175]])

In [90]:
points_t = [i for [i, j] in X_train]
price_t  = [j for [i, j] in X_train]

In [91]:
points_t[0], price_t[0]

(0.9287715297502791, 0.6281210942067972)

In [116]:
X_train = [[points_t[i],price_t[i]] for i in range(0,len(price))]

In [117]:
X_train[0]

[0.9287715297502791, 0.6281210942067972]

# Fit models

In [118]:
LR_model.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='multinomial',
          n_jobs=None, penalty='l2', random_state=0, solver='lbfgs',
          tol=0.0001, verbose=0, warm_start=False)

In [119]:
LDA_model.fit(X_train, y_train)

LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
              solver='svd', store_covariance=False, tol=0.0001)

In [120]:
KNN_model.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform')

In [121]:
GaussianNB_model.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [122]:
DTree_model.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [123]:
SVC_model.fit(X_train, y_train)

/home/reedless/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:931: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)

In [124]:
#xgb_model.fit(X_train, y_train)

In [125]:
# Save predictions for evaluation
list_of_predictions = []

LR_prediction    = LR_model.predict(X_test)
list_of_predictions.append(LR_prediction)

LDA_prediction   = LDA_model.predict(X_test)
list_of_predictions.append(LDA_prediction)

KNN_prediction   = KNN_model.predict(X_test)
list_of_predictions.append(KNN_prediction)

GNB_prediction   = GaussianNB_model.predict(X_test)
list_of_predictions.append(GNB_prediction)

DTree_prediction = DTree_model.predict(X_test)
list_of_predictions.append(DTree_prediction)

SVC_prediction   = SVC_model.predict(X_test)
list_of_predictions.append(SVC_prediction)


In [126]:
#XGB_prediction   = xgb_model.predict(X_test)
#list_of_predictions.append(XGB_prediction)


In [127]:
# Accuracy score is the simplest way to evaluate
for prediction in list_of_predictions:
    print(accuracy_score(prediction, y_test))

0.16621738711851353
0.16159622605179552
0.16313661307403485
0.31707904110907864
0.14782901704053145
0.17175315297968616


# Results of experimental scaling
### No scaling
0.598408266743686
0.5987291807066526
0.5256891627354706
0.5945893905843843
0.6076505888771221
0.46086454221623185

### Scale points
0.1792625397131029
0.1792625397131029
0.5998844709733321
0.599836333878887
0.5315779339559065
0.1421007028015789

### Scale price
0.14725137190719167
0.14277462212380862
0.15639741985173775
0.14238952536824878
0.4151824395879465
0.14238952536824878

### Scale both
0.16621738711851353
0.16159622605179552
0.16313661307403485
0.31707904110907864
0.14782901704053145
0.17175315297968616

In [72]:
y_test.value_counts()

US        12461
Italy      3724
France     2958
Spain      1631
Name: country, dtype: int64

In [73]:
# confusion_matrix for all predictions
for prediction in list_of_predictions:
    print(confusion_matrix(prediction, y_test))

[[1225 1496  328 5281]
 [1733 2228 1303 7180]
 [   0    0    0    0]
 [   0    0    0    0]]
[[1431 1798  417 6627]
 [1527 1926 1214 5834]
 [   0    0    0    0]
 [   0    0    0    0]]
[[1801 2194  583 8352]
 [1142 1515 1011 4036]
 [   0    0    0    0]
 [  15   15   37   73]]
[[1568 1955  479 7442]
 [   0    0    0    0]
 [   0    0    0    0]
 [1390 1769 1152 5019]]
[[ 2318  2974   902 10496]
 [  640   748   728  1960]
 [    0     0     0     0]
 [    0     2     1     5]]
[[ 894 1051  229 2961]
 [2064 2673 1402 9500]
 [   0    0    0    0]
 [   0    0    0    0]]
[[ 2950  3712  1607 12397]
 [    8    12    24    64]
 [    0     0     0     0]
 [    0     0     0     0]]


In [74]:
# classification_report for all predictions
for prediction in list_of_predictions:
    print(classification_report(prediction, y_test))

/home/reedless/anaconda3/lib/python3.7/site-packages/sklearn/metrics/classification.py:1145: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


              precision    recall  f1-score   support

      France       0.41      0.15      0.22      8330
       Italy       0.60      0.18      0.28     12444
       Spain       0.00      0.00      0.00         0
          US       0.00      0.00      0.00         0

   micro avg       0.17      0.17      0.17     20774
   macro avg       0.25      0.08      0.12     20774
weighted avg       0.52      0.17      0.25     20774

              precision    recall  f1-score   support

      France       0.48      0.14      0.22     10273
       Italy       0.52      0.18      0.27     10501
       Spain       0.00      0.00      0.00         0
          US       0.00      0.00      0.00         0

   micro avg       0.16      0.16      0.16     20774
   macro avg       0.25      0.08      0.12     20774
weighted avg       0.50      0.16      0.24     20774

              precision    recall  f1-score   support

      France       0.61      0.14      0.23     12930
       Italy       0.